### PR TITLE
feat: garden DTO

### PR DESCRIPTION
**Pull request by Gustav-Eikaas**
# Description

Please include a summary of the change and which issue is fixed.
List any external dependencies that are required for this change.

Completes: [AB#33460](https://statoil-proview.visualstudio.com/9949ad5e-7d15-4531-901e-296574079ee1/_workitems/edit/33460)

## Type of change

Remove options that are not relevant.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] Refactor (A code change that neither fixes a bug nor adds a feature)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist

-   [ ] My code follows the style guidelines of this project
-   [ ] I have performed a self-review of my own code.
-   [ ] I have linked my DevOps task using the AB# tag.
-   [ ] My code is easy to read.
-   [ ] Documentation and comments is added where needed.
-   [ ] My code is covered by tests.
-   [ ] UX has reviewed relevant UI contributions.

### CODE REVIEW

File: packages/garden/src/lib/classes/gardenController.ts [hunk 1]
```diff
@@ -168,10 +163,16 @@ export class GardenController<
...
 	groupData = () => {
 		this.groups.setValue(
-			this.postGroupSorting(createGarden(this), [
-				this.grouping.value.horizontalGroupingAccessor as keyof TData,
-				...(this.grouping.value.verticalGroupingKeys as (keyof TData)[]),
-			])
+			this.postGroupSorting(
+				createGarden(
+					//TODO: Investigate TContext being interpreted as Record<PropertyKey,unknown>
+					this as GardenController<TData, ExtendedFields, TCustomGroupByKeys, Record<PropertyKey, unknown>>
```
Comment (Gustav-Eikaas) [inline packages/garden/src/lib/classes/gardenController.ts:169]: fix this

File: packages/garden/src/lib/components/DefaultGardenItem/DefaultGardenItem.tsx [hunk 2]
```diff
@@ -18,15 +18,9 @@ export const DefaultGardenItem = memo(
...
+			const label = controller.getDisplayName(
 				item,
-				controller as GardenController<
-					Record<PropertyKey, unknown>,
-					never,
-					never,
-					Record<PropertyKey, unknown>,
-					Record<PropertyKey, unknown>
-				>
+				controller as GardenController<Record<PropertyKey, unknown>, never, never, Record<PropertyKey, unknown>>
```
Comment (Gustav-Eikaas) [inline packages/garden/src/lib/components/DefaultGardenItem/DefaultGardenItem.tsx:23]: fix this

File: packages/garden/src/lib/components/Garden.tsx [hunk 3]
```diff
@@ -32,8 +30,7 @@ export function Garden<
 				controller as unknown as GardenController<
 					Record<PropertyKey, unknown>,
 					string,
-					BaseRecordObject<unknown>,
-					BaseRecordObject<unknown>,
```
Comment (Gustav-Eikaas) [inline packages/garden/src/lib/components/Garden.tsx:36]: fix this

File: packages/garden/src/lib/components/Garden.tsx [hunk 4]
```diff
@@ -44,11 +41,5 @@ export function Garden<
...
 export const GardenContext = createContext<
-	GardenController<
-		Record<PropertyKey, unknown>,
-		string,
-		BaseRecordObject<unknown>,
-		BaseRecordObject<unknown>,
-		Record<PropertyKey, unknown>
-	>
+	GardenController<Record<PropertyKey, unknown>, string, Record<PropertyKey, unknown>, Record<PropertyKey, unknown>>
 >({} as any);
```
Comment (Gustav-Eikaas) [inline packages/garden/src/lib/components/Garden.tsx:45]: fix this

File: packages/garden/src/lib/components/GardenItemContainer/GardenItemContainer.tsx [hunk 5]
```diff
@@ -121,7 +121,14 @@ export const GardenItemContainer = <
 								columnExpanded={
 									expand?.expandedColumns?.[groups[virtualColumn.index].value]?.isExpanded ?? false
 								}
-								controller={controller as any}
+								controller={
+									createGardenProp(controller) as GardenProp<
+										TData,
+										TExtendedFields,
```
Comment (Gustav-Eikaas) [inline packages/garden/src/lib/components/GardenItemContainer/GardenItemContainer.tsx:127]: fix this

File: packages/garden/src/lib/utils/createGardenProp.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+	TContext extends Record<PropertyKey, unknown> = Record<PropertyKey, unknown>
+>(
+	controller: GardenController<TData, ExtendedFields, TCustomGroupByKeys, TContext>
+): GardenProp<TData, ExtendedFields, TCustomGroupByKeys, TContext> => ({
+	onClickItem: (item) => controller.clickEvents.onClickItem && controller.clickEvents.onClickItem(item, controller),
+	getDisplayName: (item: TData) =>
+		controller.getDisplayName(
+			item,
+			//TODO: Investigate why default value never breaks type inferring
+			controller as GardenController<TData, never, never, Record<PropertyKey, unknown>>
```
Comment (Gustav-Eikaas) [inline packages/garden/src/lib/utils/createGardenProp.ts:18]: this fix

Comment (Gustav-Eikaas) [reply packages/garden/src/lib/utils/createGardenProp.ts:18]: Can't fix

File: packages/garden/src/lib/types/gardenProp.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+	useData: () => TData[];
+	/** Will re-render when grouping changes */
+	useGroups: () => GardenGroup<TData>[];
+	/** Will re-render when grouping changes */
+	useCurrentGroupingKeys: () => GroupingKeys<TData>;
+	getDisplayName: (item: TData) => string;
+	getIdentifier: GetIdentifier<TData>;
+	onClickItem: (item: TData) => void;
+	useContext: () => TContext | undefined;
+	/** Will rerender when context changes */
```
Comment (jaysencpp) [inline packages/garden/src/lib/types/gardenProp.ts:23]: double comment

Comment (Gustav-Eikaas) [reply packages/garden/src/lib/types/gardenProp.ts:22]: fixed

**Review by jaysencpp — APPROVED** [review]

**Pull request merged**